src/NuGet/Microsoft.Net.Compilers.Toolset/DesktopCompilerArtifacts.targets

```diff
@@ -62,19 +62,19 @@
         
         We don't currently collect optimization data for the following assemblies.
       -->
-      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Buffers.dll"/>
+      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Buffers.dll" NeededForBuildTask="true"/>
       <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Collections.Immutable.dll"/>
-      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Memory.dll"/>
-      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Numerics.Vectors.dll"/>      
+      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Memory.dll" NeededForBuildTask="true"/>
+      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Numerics.Vectors.dll"/>
       <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Reflection.Metadata.dll"/>
-      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Runtime.CompilerServices.Unsafe.dll"/> 
-      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Text.Encoding.CodePages.dll"/>   
-      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Threading.Tasks.Extensions.dll"/>  
+      <_NoOptimizationData Include="$(ArtifactsBinDir)csi\$(Configuration)\net472\System.Runtime.CompilerServices.Unsafe.dll" NeededForBuildTask="true"/>
```

Comment:
I understand why you're doing this but it feels very flaky. Basically we need to be _really_ careful about when we take changes to our build taks that bring in new dependencies that we don't accidentally forget to give them this property. 

I wonder if we should bite the bullet for net472 and put the tools into a sub-directory below the build task. Essentially make our desktop layout match our core layout where the compiler is in `bincore`. That would make the dependencies here much clearer. 

Can you file a follow up issue to take care of this later? Don't want to change the scope of this PR but do want to deal with this before we ship Dev18.

Reply:
https://github.com/dotnet/roslyn/issues/78001